controller/controller.go

```diff
@@ -151,72 +139,62 @@ func appHandler(c handlerConfig) (http.Handler, *martini.Martini) {
+		resourceRepo:    resourceRepo,
+		clusterClient:   c.cc,
+		discoverdClient: c.dc,
+		routerc:         c.sc,
+	}
+
 	httpRouter := httprouter.New()
-	httpRouter.NotFound = http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
-		m.ServeHTTP(res, req)
-	})
 
-	getAppMiddleware := crud(httpRouter, "apps", ct.App{}, appRepo)
-	getReleaseMiddleware := crud(httpRouter, "releases", ct.Release{}, releaseRepo)
-	getProviderMiddleware := crud(httpRouter, "providers", ct.Provider{}, providerRepo)
+	crud(httpRouter, "apps", ct.App{}, appRepo)
+	crud(httpRouter, "releases", ct.Release{}, releaseRepo)
+	crud(httpRouter, "providers", ct.Provider{}, providerRepo)
 	crud(httpRouter, "artifacts", ct.Artifact{}, artifactRepo)
 	crud(httpRouter, "keys", ct.Key{}, keyRepo)
 
-	r.Put("/apps/:apps_id/formations/:releases_id", getAppMiddleware, getReleaseMiddleware, binding.Bind(ct.Formation{}), putFormation)
-	r.Get("/apps/:apps_id/formations/:releases_id", getAppMiddleware, getFormationMiddleware, getFormation)
-	r.Delete("/apps/:apps_id/formations/:releases_id", getAppMiddleware, getFormationMiddleware, deleteFormation)
-	r.Get("/apps/:apps_id/formations", getAppMiddleware, listFormations)
-
-	r.Post("/apps/:apps_id/jobs", getAppMiddleware, binding.Bind(ct.NewJob{}), runJob)
-	r.Get("/apps/:apps_id/jobs/:jobs_id", getAppMiddleware, getJob)
-	r.Put("/apps/:apps_id/jobs/:jobs_id", getAppMiddleware, binding.Bind(ct.Job{}), putJob)
-	r.Get("/apps/:apps_id/jobs", getAppMiddleware, listJobs)
-	r.Delete("/apps/:apps_id/jobs/:jobs_id", getAppMiddleware, connectHostMiddleware, killJob)
-	r.Get("/apps/:apps_id/jobs/:jobs_id/log", getAppMiddleware, connectHostMiddleware, jobLog)
-
-	r.Put("/apps/:apps_id/release", getAppMiddleware, binding.Bind(releaseID{}), setAppRelease)
-	r.Get("/apps/:apps_id/release", getAppMiddleware, getAppRelease)
-
-	r.Post("/providers/:providers_id/resources", getProviderMiddleware, binding.Bind(ct.ResourceReq{}), resourceServerMiddleware, provisionResource)
-	r.Get("/providers/:providers_id/resources", getProviderMiddleware, getProviderResources)
-	r.Get("/providers/:providers_id/resources/:resources_id", getProviderMiddleware, getResourceMiddleware, getResource)
-	r.Put("/providers/:providers_id/resources/:resources_id", getProviderMiddleware, binding.Bind(ct.Resource{}), putResource)
-	r.Get("/apps/:apps_id/resources", getAppMiddleware, getAppResources)
-
-	r.Post("/apps/:apps_id/routes", getAppMiddleware, binding.Bind(router.Route{}), createRoute)
-	r.Get("/apps/:apps_id/routes", getAppMiddleware, getRouteList)
-	r.Get("/apps/:apps_id/routes/:routes_type/:routes_id", getAppMiddleware, getRouteMiddleware, getRoute)
-	r.Delete("/apps/:apps_id/routes/:routes_type/:routes_id", getAppMiddleware, getRouteMiddleware, deleteRoute)
-
-	r.Get("/formations", getFormations)
-
-	return muxHandler(httpRouter, c.key), m
+	httpRouter.PUT("/apps/:apps_id/formations/:releases_id", httphelper.WrapHandler(api.PutFormation))
```

Comment:
As discussed we will wrap httprouter to eliminate wrapping each individual handler (and enable adding things like route grouping) after this PR has landed.